### PR TITLE
[102X] add b-taggers WPs for AK4 jets, rename Jet::btag_DeepFlavour() to Jet::btag_DeepJet(), specify era of process 

### PR DESCRIPTION
**Pull request by karavdin**
add working points for b-taggers for 2016, 2017 and 2018;
change list of discriminators to enable storage of DeepCSV for AK4PUPPI in 2016v2
rename `Jet::btag_DeepFlavour()` to `Jet::btag_DeepJet()`

Check with Z'->ttbar selection for tight WP of each b-tagger on AK4PUPPI for samples of different years:
b_taggers_zprime

### CODE REVIEW

**Comment by raggleton** [comment]
Hm so here you're now rerunning the normal btagging discriminators when you add in the deep* discriminators. However it seems to change the values for discriminators already stored in the jets in miniaod (looking at jetAk4CHS and jetsAk4Puppi). I would have thought that rerunning the discriminators should give the same answer as the originally stored value? e.g. https://uhh2-integration.web.cern.ch/UHH2integration/test1163/data_2018_SingleMu_figs/jetsAk4Puppi.btag_combinedSecondaryVertexMVA.pdf

And for the jetsAk8*Substructure* collections, there should be 0 difference at all, since those b discriminators are already calculated when `addJetCollection()` is run in `add_fatjets_subjets()`. Most look near identical (probably a very subtle floating point difference), but some are visibly different, which is a tad worrying: https://uhh2-integration.web.cern.ch/UHH2integration/test1163/mc_2016v2_TTbar_figs/jetsAk8PuppiSubstructure_SoftDropPuppi.btag_DeepCSV.pdf

**Comment by raggleton** [comment]
Also can you update the title to also mention renaming `Jet::btag_DeepFlavour()` to `Jet::btag_DeepJet()` just for future reference? 
I'm also now wondering if all the other `Jet::btag_DeepFlavour_*()` should be changed to `btag_DeepJet_*` for consistency...

**Comment by karavdin** [comment]
Most probably we can ignore combinedSecondaryVertexMVA, which is not supported beyond 2016v2. 
CSV change seems to be related to migration of few events from one bin to another and I would not worry to much about it. 
DeepCSV changes are spectacular, even for 2018, where we suppose to use the same settings as in miniAOD, but result is different for both CHS and PUPPI
[jetsAk4CHS.btag_DeepCSV.pdf](https://github.com/UHH2/UHH2/files/2892850/jetsAk4CHS.btag_DeepCSV.pdf)
[jetsAk4Puppi.btag_DeepCSV.pdf](https://github.com/UHH2/UHH2/files/2892852/jetsAk4Puppi.btag_DeepCSV.pdf)

**Comment by karavdin** [comment]
> Also can you update the title to also mention renaming `Jet::btag_DeepFlavour()` to `Jet::btag_DeepJet()` just for future reference?
> I'm also now wondering if all the other `Jet::btag_DeepFlavour_*()` should be changed to `btag_DeepJet_*` for consistency...

Added info to the title. For getters of the discriminant inputs, I would be more in favour to leave the same names as variables, so setters/getters are consistent.

**Comment by karavdin** [comment]
Outcome of DeepCSV test on AK4PUPPI with and without re-running it on 2018 ttbar all hadronic:
btaggerstest_10gev
btaggerstest_30gev
btaggerstest_100gev

The difference is presented in jets with relatively high pt, thus hypothesis that the issue is related to low pt jet is excluded.

**Comment by karavdin** [comment]
Split DeepCSV into probb and probbb:
btaggerstest_100gev_probb
btaggerstest_100gev_probbb

**Comment by karavdin** [comment]
Asked experts on HN: https://hypernews.cern.ch/HyperNews/CMS/get/btag/1644.html

**Comment by karavdin** [comment]
As suggested by experts, adding era to config. Test on 2018 TTbar looks promissing
btaggerstest_100gev
btaggerstest_10gev

File: core/python/ntuple_generator.py [hunk 1]
```diff
@@ -68,7 +68,14 @@ def generate_process(year, useData=True, isDebug=False, fatjet_ptmin=120.):
     if (year=="2016v2" or year=="2016v3"):
         met_sources_GL.extend(['slMETsCHS'])
 
-
+    #try eras for correct b-tagging
+    from Configuration.StandardSequences.Eras import eras
+    if year == "2018":
+        Process = cms.Process("DUMMY", eras.Run2_2018)
```
Comment (raggleton) [inline core/python/ntuple_generator.py]: these need to be `process = ...`?

Comment (raggleton) [reply core/python/ntuple_generator.py]: would then need to remove the `process = cms.Process...` line at L171

Comment (karavdin) [reply core/python/ntuple_generator.py]: hm, not sure why it worked since DUMMY process is indeed not used anywhere. Will fix it in the next commit.

**Comment by karavdin** [comment]
Ok, after specifying era, changes still related to taggers, so nothing else is touched by it.
In 2016 v2 we're adding DeepCSV for PUPPI (expected changes)
In 2016 v3 / 2017 there are some changes in  SingleElectron and SingleMuon, e.g SingleMuon (ak4PUPPI):
https://uhh2-integration.web.cern.ch/UHH2integration/test1163/data_2016v3_SingleMu_figs/jetsAk4Puppi.btag_DeepCSV.pdf
Are eras for 2016v3/2017 lepton datasets wrong?
In that respect 2018 data looks great, some hist with different RMS most probably related to migration of few events from one bin to another.
For MC, again 2016v3/2017 are different for AK4PUPPI 
https://uhh2-integration.web.cern.ch/UHH2integration/test1163/mc_2016v3_TTbar_figs/jetsAk4Puppi.btag_DeepCSV.pdf
https://uhh2-integration.web.cern.ch/UHH2integration/test1163/mc_2017_TTHadronic_figs/jetsAk4Puppi.btag_DeepCSV.pdf
https://uhh2-integration.web.cern.ch/UHH2integration/test1163/mc_2017_TTSemiLeptonic_figs/jetsAk4Puppi.btag_DeepCSV.pdf

Based on this, it seems like eras for 2016v3 and 2017 are not appropriate, but for some reason it's visible only for PUPPI and CHS is less affected.

**Comment by raggleton** [comment]
So doing some digging it seems that:

- 2017v2 MC needs `eras.Run2_2017, eras.run2_miniAOD_94XFall17`
- 2017v1 MC needs `eras.Run2_2017`
- 2016v3 MC needs `eras.Run2_2016, eras.run2_miniAOD_80XLegacy`
- 2016v2 MC needs `eras.Run2_2016`

The mc_2017_TTHadronic is 2017v1 MC, whilst the mc_2017_TTLeptonic is 2017v2 MC.

As for data, I am finding it harder to definitively find out which to use for 2017. Our tests run over `31Mar2018` datasets, which I _think_ also uses `eras.Run2_2017, eras.run2_miniAOD_94XFall17`. So really it's 2017v2.

To find out for MC: I had to get prepid from DAS, then lookup in McM for the cmsDriver command.

To find out for data: I had to go to DAS, find the MiniAOD dataset. Then find its RAW parent. Then look at its `config` key, and open the output configs, one of which should be the miniaod. Except they look more like DQM than actual miniaod, so I'm not 100% sure at all.

**Comment by karavdin** [comment]
Merging to start processing of 2018 ntuples for JERC. 2017 needs further investigation.

**Comment by karavdin** [comment]
2017 TTbar semileptonic: 
/store/mc/RunIIFall17MiniAODv2/TTToSemiLeptonic_TuneCP5_13TeV-powheg-pythia8/MINIAODSIM/PU2017_12Apr2018_94X_mc2017_realistic_v14-v2/30000/D221F074-FF58-E811-958D-509A4C78138B.root
Eras: eras.Run2_2017, eras.run2_miniAOD_94XFall17
btaggerstest_10gev
btaggerstest_100gev

**Comment by karavdin** [comment]
2017 TTbar semileptonic:
/store/mc/RunIIFall17MiniAODv2/TTToSemiLeptonic_TuneCP5_13TeV-powheg-pythia8/MINIAODSIM/PU2017_12Apr2018_94X_mc2017_realistic_v14-v2/30000/D221F074-FF58-E811-958D-509A4C78138B.root
Eras: eras.Run2_2017
btaggerstest_10gev
btaggerstest_100gev

**Pull request merged**